### DISCUSSION TITLE
Pyright issue: Arguments missing for parameters

### DISCUSSION THREAD

**Discussion by kbakk** [pydantic/pydantic#7379] [Question] [answered]
Hello! 👋 

Given these Pydantic models

```py
from typing import Optional

from pydantic import BaseModel, Field


class NestedModel(BaseModel):
    attribute_a: Optional[int] = Field(1)
    attribute_b: Optional[int] = Field(2)
    attribute_c: Optional[str] = Field("3")


class MainModel(BaseModel):
    nested: NestedModel = NestedModel()
```

I'm getting an error from Pyright:

```
.../repro_pydantic.py
  .../repro_pydantic.py:13:27 - error: Arguments missing for parameters "attribute_a", "attribute_b", "attribute_c" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

I am using Field to document (description, examples etc.) so replacing them with literal values is not an option here.

Does anyone have a suggestion how to tell Pyright that these are not required? Or am I maybe doing something strange here?

**Comment by kbakk** [comment]
This seems to work (not sure if there are any typing nuances around this to be aware of though):

```py
from typing import Optional, Annotated

from pydantic import BaseModel, Field


class NestedModel(BaseModel):
    attribute_a: Annotated[Optional[int], Field(description="A")] = 1
    attribute_b: Annotated[Optional[int], Field(description="B")] = 2
    attribute_c: Annotated[Optional[str], Field(description="C")] = "3"


class MainModel(BaseModel):
    nested: NestedModel = NestedModel()


if __name__ == "__main__":
    print(MainModel().model_json_schema())
# {'$defs': {'NestedModel': {'properties': {'attribute_a': {'anyOf': [{'type': 'integer'}, {'type': 'null'}], 'default': 1, 'description': 'A', 'title': 'Attribute A'}, 'attribute_b': {'anyOf': [{'type': 'integer'}, {'type': 'null'}], 'default': 2, 'description': 'B', 'title': 'Attribute B'}, 'attribute_c': {'anyOf': [{'type': 'string'}, {'type': 'null'}], 'default': '3', 'description': 'C', 'title': 'Attribute C'}}, 'title': 'NestedModel', 'type': 'object'}}, 'properties': {'nested': {'allOf': [{'$ref': '#/$defs/NestedModel'}], 'default': {'attribute_a': 1, 'attribute_b': 2, 'attribute_c': '3'}}}, 'title': 'MainModel', 'type': 'object'}
```

Worked it out from the example here: https://docs.pydantic.dev/latest/api/fields/#pydantic.fields.FieldInfo.from_annotated_attribute

**Reply by Viicos** [reply, marked as answer]
`Annotated` is a great way to deal with this issue, as the specified default argument (e.g. ` = 1`) is the "real" default value, whereas using ` = Field(...)` can be counterintuitive, especially if you don't specify a default value with `Field`. This is mostly why [FastAPI recommends the usage of `Annotated`](https://fastapi.tiangolo.com/tutorial/query-params-str-validations/#advantages-of-annotated).

[PEP 681 states](https://peps.python.org/pep-0681/#field-specifier-parameters) field parameters must be keyword arguments. I think Pydantic had the `Field` function before supporting PEP 681, and this why `default` isn't a kw-only argument:

https://github.com/pydantic/pydantic/blob/9c03680ff08fceabfba0db6402e12ca8970bfdcf/pydantic/fields.py#L666-L671

The issue you encountered has been reported here: https://github.com/pydantic/pydantic/issues/3617. While the right thing for Pydantic would be to make `default` a kw-only argument to fully comply with PEP 681, I don't think it is doable in practice, considering how much of a breaking change it is, + I feel using `Annotated` should be encouraged nowadays.

**Reply by maximvl** [reply]
Sorry I don't fully understand the answer, is annotation required even for arguments that are supposed to be optional without any default value? It doesn't make much sense to me

![Screenshot 2024-04-03 at 17 37 16](https://github.com/pydantic/pydantic/assets/949890/d9180fda-8a19-4224-a823-a4632ca58540)

**Reply by Viicos** [reply]
I'm guessing your model is defined as:

```python
class MyModel(BaseModel):
    id: int
    name: Optional[str]
```

In that case, yes, `name` is required. `Optional` doesn't mean the field doesn't have to be provided. However, it was (wrongfully) the case in V1, see [this section of the docs](https://docs.pydantic.dev/latest/migration/#required-optional-and-nullable-fields).

**Comment by TimChild** [comment]
I prefer to just specify the `default` as a kw parameter:
```python
class Example(BaseModel):
    has_a_default: int = Field(default=0, description="a description")

example = Example()  # No pyright error since `default` specified as a keyword
```

As also suggested here https://github.com/microsoft/pyright/issues/6915

**Comment by baggiponte** [comment]
I am using `default_factory` and still get this error: 

```python
from datetime import date
from typing import Annotated

from pydantic import BaseModel, Field


class MyModel(BaseModel):
    current_date: Annotated[date, Field(default_factory=date.today)]


mod = MyModel()
```

<img width="680" alt="image" src="https://github.com/user-attachments/assets/efd24232-35fd-4606-b53c-ce422643e152" />

However if I do 

```diff
from datetime import date
from typing import Annotated

from pydantic import BaseModel, Field


class MyModel(BaseModel):
-    current_date: Annotated[date, Field(default_factory=date.today)]
+   current_date: Annotated[date, Field(default_factory=date.today)] = ...

mod = MyModel()
```

The first error disappears, but I get a different one:

Of course I get a different error:

<img width="1190" alt="image" src="https://github.com/user-attachments/assets/961ff508-e39b-4dee-9d97-4088486af5f8" />

EDIT:

I don't get any error if I do this:

```python
from datetime import date

from pydantic import BaseModel, Field


class MyModel(BaseModel):
    current_date: date = Field(default_factory=date.today)


mod = MyModel()
```

<img width="615" alt="image" src="https://github.com/user-attachments/assets/2d2ed100-4a01-4f15-89a8-b09a849a456d" />

**Reply by Viicos** [reply]
See https://docs.pydantic.dev/latest/concepts/fields/#the-annotated-pattern:

![image](https://github.com/user-attachments/assets/aafcfe95-7a7d-42e3-a773-ddb75d4cd725)

**Reply by baggiponte** [reply]
Thanks for the reply!
